### PR TITLE
Add smoke tests to 'Release Version' workflow [HZ-2748]

### PR DESCRIPTION
**Pull request by sumnerib**
Adds module to build and run a simple 'smoke test' during the release
deployment pipline. Runs separate steps for OS and EE jars.

~~Requires a new repository secret (`HZ_ENTERPRISE_LICENSE`) to run the HZ
EE service container.~~

### CODE REVIEW

File: .github/workflows/deploy-release.yml [hunk 1]
```diff
@@ -21,6 +21,17 @@ jobs:
   do-release:
     if: github.repository_owner == 'hazelcast'
     runs-on: ubuntu-latest
+    services:
+      hazelcast:
+        image: hazelcast/hazelcast
```
Comment (kwart) [inline .github/workflows/deploy-release.yml:26]: This should be faster:
```suggestion
        image: hazelcast/hazelcast:latest-slim
```

File: .github/workflows/deploy-release.yml [hunk 2]
```diff
@@ -21,6 +21,17 @@ jobs:
...
+        ports:
+          - 6701:5701
+      hazelcast-ee:
+        image: hazelcast/hazelcast-enterprise
```
Comment (kwart) [inline .github/workflows/deploy-release.yml:30]:
```suggestion
        image: hazelcast/hazelcast-enterprise:latest-slim
```

**Review by kwart — CHANGES_REQUESTED** [review]
I like the improvement, just left some comments.
I didn't know about the possibility of using container services, so happy to learn sth new :+1:

**Comment by kwart** [comment]
BTW When you have EE and OS hazelcast members as services, don't they try to connect one to the other?

File: .github/workflows/deploy-release.yml [hunk 3]
```diff
@@ -47,6 +58,18 @@ jobs:
           git tag v${{ github.event.inputs.release-version }}
           mvn clean install -V -B
 
+      - name: Build and run smoke test
+        timeout-minutes: 2
```
Comment (nishaatr) [inline .github/workflows/deploy-release.yml:62]: Is 2mins enough? 
Is timeout needed as I would expect calls to HZ will or should fail
Trying to understand the reason behind the timeout?

File: smoke-test/src/main/java/com/hazelcast/smoke/test/SmokeTestDriver.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+                        + "  'keyFormat'='integer', "
+                        + "  'valueFormat'='varchar'"
+                        + ")");
+
+                int insertCount = s.executeUpdate(
+                        "INSERT INTO map "
+                        + "SELECT v, 'name-' || v "
+                        + "FROM TABLE(generate_series(0, 9))"
+                );
+                assert insertCount == 0;
```
[the remark this answers is not in the document]
Comment (nishaatr) [reply smoke-test/src/main/java/com/hazelcast/smoke/test/SmokeTestDriver.java:37]: I agree with Josef considering try() is in use

[the remark this answers is not in the document]
Comment (nishaatr) [reply smoke-test/src/main/java/com/hazelcast/smoke/test/SmokeTestDriver.java:37]: Any reason for the top smoke-test folder?

We have this test folder https://github.com/hazelcast/hazelcast-jdbc/tree/360f04b177e6592f3089f5322943c321c88c2f72/hazelcast-jdbc-core/src/test/java/com/hazelcast/jdbc

[the remark this answers is not in the document]
Comment (nishaatr) [reply smoke-test/src/main/java/com/hazelcast/smoke/test/SmokeTestDriver.java:37]: Normally test files end with *Test.java
Not a issue but perhaps just SmokeTest.java?

[the remark this answers is not in the document]
Comment (kwart) [reply smoke-test/src/main/java/com/hazelcast/smoke/test/SmokeTestDriver.java:37]: I have no strong opinion on the location (existing or new module) - both sounds OK.
WRT the class name, I like the original `SmokeTestDriver` more. The `*Test` ones are usually unit tests you want to be executed by Maven automatically.

[the remark this answers is not in the document]
Comment (nishaatr) [reply smoke-test/src/main/java/com/hazelcast/smoke/test/SmokeTestDriver.java:37]: Lets keep as is 
I didn't realize this is a module (I see a POM.xml). Tbh I struggle a little with reviewing GH PRs as not a huge fan of the interface :)

[the remark this answers is not in the document]
Comment (nishaatr) [reply smoke-test/src/main/java/com/hazelcast/smoke/test/SmokeTestDriver.java:37]: I presume SmokeTestDriver.java will contain all future new tests which might proliferate the file e.g. you might want to separate or group smoke tests - hence, thought 'Driver' might not ideal 

Also, there is no ability to disable some tests apart from commenting them out like with Unit tests

**Just adding some thoughts here and no action needed.**

**Comment by sumnerib** [comment]
> BTW When you have EE and OS hazelcast members as services, don't they try to connect one to the other?

@kwart I don't have the logs from the runs on my fork, but I think they do since they have a shared network. I tested in my local docker, and indeed they attempt and fail to join via multicast:

```
2023-08-02 11:37:04,939 [ WARN] [hz.dazzling_gagarin.MulticastThread] [c.h.i.c.i.MulticastService]: [172.17.0.2]:5701 [smoke-test] [5.3.1] Received a JoinRequest with an incompatible binary-format. An old version of Hazelcast may be using the same multicast discovery port. Are you running multiple Hazelcast clusters on this host? (This message will be suppressed for 60 seconds).
```

Even though it didn't cause any issue, I will go ahead and add some config to disable multicast and add cluster names, just to be safe. Good catch!

```
          HZ_CLUSTERNAME: smoke-test
          HZ_NETWORK_JOIN_MULTICAST_ENABLED: false
          HZ_NETWORK_JOIN_TCPIP_ENABLED: true
```

File: smoke-test/src/main/java/com/hazelcast/smoke/test/SmokeTestDriver.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sumnerib) [reply smoke-test/src/main/java/com/hazelcast/smoke/test/SmokeTestDriver.java:37]: I meant to enable assertions (-ea) in my initial commit. Good catch!

@nishaatr I don't expect we should need to add any new tests to `SmokeTestDriver`. The intent is to encapsulate the most basic DDL and CRUD operations used in SQL. Basically, we want to verify that the artifact generated by maven works (in some basic sense) and the needed/expected dependencies are included in the final jar. New feature testing should continue to be in unit/integration tests.

With this in mind, if you think it should be cleaned up and broken into multiple constructs, I'm happy to make those changes as well.

File: .github/workflows/deploy-release.yml (hunk 3 above)
Comment (sumnerib) [reply .github/workflows/deploy-release.yml:71]: Good question. At first I thought the timeout would be necessary, but it seems redundant and unnecessary thinking about it more. Will remove.

**Review by nishaatr — APPROVED** [review]

**Pull request merged**